File: labs/lab1/precision.cpp
#include <iostream>
#include <limits>

int main() {
    int k = 0;
    float a = 1;
    double b = 1;
    long double c = 1;

    std::cout << "Float" << std::endl;
    while(a != 0){
        a /= 2;
        k++;
    }
    std::cout << "zero = 2^-" << k << std::endl;
    k = 0;
    a = 1;
    while(a < std::numeric_limits<float>::max()){
        a *= 2;
        k++;
    }
    std::cout << "infinity = 2^" << k << std::endl;
    k = 0;
    a = 1;
    while(1 + a > 1){
        a /= 2;
        k++;
    }
    std::cout << "epsilon = 2^-" << k << std::endl;

    std::cout << "Double" << std::endl;
    k = 0;
    while(b != 0){
        b /= 2;
        k++;
    }
    std::cout << "zero = 2^-" << k << std::endl;
    k = 0;
    b = 1;
    while(b < std::numeric_limits<double>::max()){
        b *= 2;
        k++;
    }
    std::cout << "infinity = 2^" << k << std::endl;
    k = 0;
    b = 1;
    while(1 + b > 1){
        b /= 2;
        k++;
    }
    std::cout << "epsilon = 2^-" << k << std::endl;

    std::cout << "Long Double" << std::endl;
    k = 0;
    while(c != 0){
        c /= 2;
        k++;
    }
    std::cout << "zero = 2^-" << k << std::endl;
    k = 0;
    c = 1;
    while(c < std::numeric_limits<long double>::max()){
        c *= 2;
        k++;
    }
    std::cout << "infinity = 2^" << k << std::endl;
    k = 0;
    c = 1;
    while(1 + c > 1){
        c /= 2;
        k++;
    }
    std::cout << "epsilon = 2^-" << k << std::endl;

    return 0;
}

File: labs/lab1/src/precision.py
#!/usr/bin/env python
# coding: utf-8

import numpy as np
import warnings

displayer = {
    'float': np.single,
    'double': np.double,
    'long double': np.longdouble,
}


def experiment(type_name):
    k: int = 0
    num_type = displayer[type_name]
    num = num_type(1)
    while num != 0:
        num = num_type(num / 2)
        k += 1
    print(type_name, "zero is 2^-" + str(k))
    k = 0
    num = num_type(1)
    while num != np.inf:
        num = num_type(num * 2)
        k += 1
    print(type_name, "infinity is 2^" + str(k))
    k = 0
    num = num_type(1)
    while num_type(1.) + num > num_type(1.):
        num = num_type(num / 2)
        k += 1
    print(type_name, "epsilon is 2^-" + str(k))


with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    for typename in displayer.keys():
        print()
        experiment(typename)
